grid2op/multi_agent/multiAgentEnv.py

```diff
@@ -0,0 +1,1070 @@
+        else:
+            raise NotImplementedError("Local observations are not available yet !")
+        #TODO Local observations
+        #for agent in self.agents: 
+        #    _cls_agent_action_space = ObservationSpace.init_grid(gridobj=self._subgrids_cls['observation'][agent], extra_name=agent)
+        #    self.observation_spaces[agent] = _cls_agent_action_space(
+        #        gridobj = self._subgrids_cls['observation'][agent],
+        #        env = self._cent_env,
+        #        rewardClass=self._cent_env._rewardClass,
+        #        observationClass=self._cent_env._observationClass,
+        #        actionClass=self._cent_env._actionClass,
+        #        #TODO following parameters
+        #        with_forecast=True,
+        #        kwargs_observation=None,
+        #    )
+    
+    def _build_action_spaces(self):
+        """Build action spaces from given domains for each agent
+        The action class is the same as 
+        """
+        for agent in self.agents: 
+            _cls_agent_action_space = ActionSpace.init_grid(gridobj=self._subgrids_cls['action'][agent], extra_name=agent)
+            self.action_spaces[agent] = _cls_agent_action_space(
+                gridobj=self._subgrids_cls['action'][agent],
+                actionClass=self._cent_env._actionClass,
+                legal_action=self._cent_env._game_rules.legal_action,
+            )
+    
+    
+    def _update_observations(self, observation : BaseObservation):
+        """Update agents' observations from the global observation given by the self._cent_env
+
+        Args:
+            observation (BaseObservation): _description_
+        """
+        if self.global_obs:
+            for agent in self.agents:
+                self.observations[agent] = observation.copy()
+        else:
+            #TODO
+            raise NotImplementedError("Local observations are not available yet !")
+    
+    def _verify_domains(self, domains : MADict) :
+        """It verifies if substation ids are valid
+
+        Args:
+            domains (dict): 
+                - key : agents' names
+                - value : list of substation ids
+        """
+        sum_subs = 0
+        for agent in domains.keys():
+            if not (isinstance(domains[agent], list)
+                    or isinstance(domains[agent], set)
+                    or isinstance(domains[agent], np.ndarray)
+            ):
+                raise DomainException("Agent id {agent} : The domain must be a list or a set of substation indices")
+            
+            if len(domains[agent]) == 0:
+                raise DomainException(f"Agent id {agent} : The domain is empty !")
+            
+            for sub_id in domains[agent] : 
+                if not (isinstance(sub_id, int) 
+                        or isinstance(sub_id, dt_int) 
+                        or isinstance(sub_id, np.int32) 
+                        or isinstance(sub_id, np.int64)
+                ):
```

Comment:
not a fan of this to be honest

Reply:
what do you suggest instead?

Reply:
what if someone uses an int type not listed there ? Do you pretend to know all python framework and all their int type ?